compiler/luci/import/src/Nodes/CircleSVDF.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Import/Nodes/CircleSVDF.h"
+
+#include <luci/IR/Nodes/CircleSVDF.h>
+
+#include <loco.h>
+
+#include <cassert>
+
+namespace luci
+{
+bool CircleSVDFBuilder::validate(const ValidateArgs &args) const
+{
+  const auto &inputs = args.op.inputs;
+
+  if (inputs.size() != args.op.inputs.size())
+    return false;
+
+  return true;
+}
+
+/**
+ * @brief  SVDF Node builder
+ *
+ * @note
+ */
+
+CircleNode *CircleSVDFBuilder::build(const circle::OperatorT &op,
+                                     GraphBuilderContext *context) const
+{
+  assert(context != nullptr);
+
+  auto graph = context->graph();
+  const std::vector<int32_t> &inputs = op.inputs;
+  const std::vector<int32_t> &outputs = op.outputs;
+  const auto tensors = context->reader()->tensors();
+  const auto opcodes = context->reader()->opcodes();
+  int32_t scratch_index = 0;
+
+  std::vector<CircleNode *> input_nodes;
+  for (const int32_t input_tensor_index : inputs)
+  {
+    auto input_node = context->nodefinder()->node(input_tensor_index);
+    if (input_node != nullptr)
+      input_nodes.push_back(input_node);
+    else
+      scratch_index = input_tensor_index;
+  }
+
+  auto svdf_scratch_tensor = tensors.at(scratch_index);
+  auto *svdf_scratch_node = graph->nodes()->create<CircleDanglingNode>();
+  copy_tensor_attributes(svdf_scratch_tensor, svdf_scratch_node);
```

Comment:
I'm not sure of this implementation. Please explain more about this.
Whatg is `svdf_scratch_tensor`?

Reply:
`SVDF` has one specific input tensor, that doesn't store any data (its buffer is empty), but is used in the kernel (as a scratch tensor) and stores quantization parameters, type and shape.

![svdf_inputs](https://user-images.githubusercontent.com/43214667/148922802-624e6d13-e9cb-4dcd-bfcc-d279a3d79531.png)

In order to import and save tensor parameters `GraphBuilderBase` is used. Otherwise, this specific input tensor is imported as null when using `GraphBuilder`. 

This part of code allows us not to lose the scratch tensor index.
```c++
    auto input_node = context->nodefinder()->node(input_tensor_index);
    if (input_node != nullptr)
      input_nodes.push_back(input_node);
    else
      scratch_index = input_tensor_index;
```
 `context->nodefinder()->node(input_tensor_index);` - return null for this input scratch tensor due to data buffer of this tensor is null.

After we save tensor index for this scratch tensor we can manually create a new dangling node that will simply store tensor attributes.
```c++
 auto svdf_scratch_tensor = tensors.at(scratch_index);
  auto *svdf_scratch_node = graph->nodes()->create<CircleDanglingNode>();
  copy_tensor_attributes(svdf_scratch_tensor, svdf_scratch_node);
```

Reply:
Thanks for the explanation. I think I need more understanding of this Op.
(1) there's `kind: Variable` which I think is new in compiler frontend.
can you please share what you know for this?
(2) For `CircleDanglingNode`: I'm not certain of introducing this Op for it's purpose and it's name; I need to understand more of this Op. If this `CircleDanglingNode` has specific purpose only for SVDF then we need to name this to reflect that purpose. Or we can consider reusing existing virtual node (like CircleOutputDummy)

Reply:
Q) How can I create this op with TensorFlow or pytorch?

Reply:
Q) about the word `scratch`; is this referenced or your invention? if referenced can you please share some URLs?